### ISSUE TITLE
Feature:  Add specific NEW key under wild-card selected parents, deliberately NOT CHANGING pre-existing values of the same key

### ISSUE THREAD

**Issue by dwapstra** [wwkimball/yamlpath#107]
**Is your feature request related to a problem? Please describe.**
In an existing yaml file, keys may be missing for some of the items matched in a path. I would like to be able to set the values for the *missing* keys.

e.g.
```
devices:
  R1:
    os: ios
  R2:
    # os missing
```

**Describe the solution you'd like**
A way to select non-existing keys and set their value, e.g.

```
yaml-set -g "/devices/*/os" -a generic --mustnotexist testbed.yaml
```

**Comment by wwkimball** [comment]
I've changed the subject of this issue to be more precise about the issue.  To wit, yaml-set will indeed build any and all necessary YAML structure to set a novel value where keys do not already exist _as long as the new structure is compatible with the old_.  What you've discovered is that yaml-set will actively refuse to convert scalar nodes into complex nodes.  In other words, yaml-set has _correctly_ detected that `/devices/R2` is a scalar node with the value `null` but the user request is to convert it into a map (Hash) and set another scalar value to some new child node of it.  Under most use-cases, it is a potential document corruption to convert a scalar node to any other complex type, or even from one complex type to another.  Were `R2` already a map (Hash), the operation would succeed.  Take for example:

```
# Use cat to experiment without actually changing testbed.yaml
# cat testbed.yaml | yaml-set --change=/devices/R3/os/sub/type --value=original
---
devices:
  R1:
    os: ios
  R2:
  R3:
    os:
      sub:
        type: original
    # os missing
```

Note the following:

1. An entirely new complex map (Hash) was created under `/devices` -- already a map (Hash) -- for the new `R3` node.
2. The new `R3` node was itself automatically built up to hold the specified end-value, `original` under a sub-child node named `type`, itself a child of `sub`, itself a child of `os`.
3. The end-of-document comment, `#os missing` correctly retained its position at the end of the document; it is not a child of `/devices/R2` because `/devices/R2` is a scalar `null` value, which properly terminates on the line above the unlinked comment.

This behavior is both intentional and necessary.  It is impossible to merge complex and scalar data together without destroying existing data (`null` is a scalar value which should be replaced only with other scalar values).

However, I discovered a real bug while experimenting with this.  Both `yaml-set` and `yaml-merge` are improperly refusing to allow users to change a scalar `null` to any other scalar value.  _That_ is unintentional and I'll take this issue as an opportunity to track and fix that behavior.

Your intention with the demonstrated YAML is a reasonable use-case and you can use the tools as they are today to solve your use-case for specific nodes.  If I had to convert a scalar node into a complex node, I would first delete the scalar node, then set it.  This is "safe" as far as the tools are concerned because the operations are both discrete and deliberate.  For example:

```
# Use cat to experiment without actually changing testbed.yaml
# cat testbed.yaml | yaml-set --change=/devices/R2 --delete | yaml-set --change=/devices/R2/os --value=generic
---
devices:
  R1:
    os: ios
    # os missing

  R2:
    os: generic
```

Note that the old _scalar_ `R2` node was deleted before an entirely new `R2` map (Hash) was added to the parent, `devices`.  [Aside:  I'm aware the new node was added after the end-of-document comment; this is something I already have on my radar to fix for a future release.]

All behaviors are open to improvement.  If deleting a scalar node to convert it to a complex node is too onerous for a wide use-case, I'm happy to explore this.

**Comment by dwapstra** [comment]
Thanks for the swift and elaborate analysis. I realize I was not accurate enough with my description.

The yaml structure has more values for each node than I showed in the example, e.g.

```
devices:
  R1:
    os: ios
    type: router
    platform: asr1k

  R2:
    type: switch
    platform: cat3k
```

So the R2 node is actually a hash value, where the `os` key is missing.

**Comment by wwkimball** [comment]
Running your test command against your sample data correctly sets `os: generic` in both `R1` and `R2`, as:

```
$ yaml-set --version
yaml-set 3.4.0

$ cat testbed.yaml
---
devices:
  R1:
    os: ios
    type: router
    platform: asr1k

  R2:
    type: switch
    platform: cat3k

$ cat testbed.yaml | yaml-set -g "/devices/*/os" -a generic
---
devices:
  R1:
    os: generic
    type: router
    platform: asr1k

  R2:
    type: switch
    platform: cat3k

    os: generic
```

Would you please elaborate a bit on how you are able to cause a failure or are otherwise unable to accomplish your goal?

**Comment by wwkimball** [comment]
Sorry I failed to fully comprehend your original intent with this request; it's my fault for being in a bug-squashing mode when your ticket originally came across.  I have again updated the title to be clear about what you are seeking.  I believe I understand, and I agree this is a useful capability.

**Comment by wwkimball** [comment]
I know it has been awhile, so I'm dropping a note here to say that I'm actively working on this feature request.  I had to push this to a future 3.5.0 release because I needed to get 3.4.1 release out, first.  This was pushed because I imagine this feature will require a more substantial change to the yamlpath processor than other bug fixes that needed to go out sooner.

Now that 3.4.1 is live, I'm working on this.  For interest, I have already made a change that makes it possible to match nodes with values less than one character long.  This only _almost_ provides the capability you seek, as titled here because it'll match nodes with zero-length values **and** non-existent nodes.  Since you only want to match non-existent nodes, my change amounts to a near-miss.

For example, this output shows what is now possible:

```
cat testbed.yaml; cat testbed.yaml | yaml-set --change='/devices/*/[os!=~/.+/]/os' --value=generic | yaml-diff .\testbed.yaml -
---
devices:
  R1:
    os: ios
    type: router
    platform: asr1k
  R2:
    type: switch
    platform: cat3k
  R3:
    type: access-point
    platform: wrt
    os:
  R4:
    type: tablet
    os: null
    platform: java
  R5:
    type: mobile
    os: ""
    platform: objective-c
  R6:
    type: tablet
    os: ''
    platform: objective-c
a devices.R2.os
> "generic"

c devices.R5.os
< ""
---
> "generic"

c devices.R6.os
< ""
---
> "generic"
```

The command now adds the missing `os` child to parents that don't have one.  However, it _also_ (correctly, per the RegEx) changes the zero-length values of `os`.

I'll keep at it.

**Comment by wwkimball** [comment]
After drawing more inspiration from XPath, I have opted to start supporting some keyword search capabilities.  The first of these, `[has_child(name)]` precisely meets the requirements of your feature request.  Here's the capability in action:

```
$ cat testbed.yaml; cat testbed.yaml | yaml-set --change='/devices/*[!has_child(os)]/os' --value=generic | yaml-diff testbed.yaml -
---
devices:
  R1:
    os: ios
    type: router
    platform: asr1k
  R2:
    type: switch
    platform: cat3k
  R3:
    type: access-point
    platform: wrt
    os:
  R4:
    type: tablet
    os: null
    platform: java
  R5:
    type: mobile
    os: ""
    platform: objective-c
  R6:
    type: tablet
    os: ''
    platform: objective-c

a devices.R2.os
> "generic"
```

As the `yaml-diff` shows, the only change is precisely what you seek:  _only_ the previously non-existent `os` child node is added and set to the provided value.  All other pre-existing `os` child nodes -- with or without a value -- are left untouched.

This is in a proof-of-concept state right now in the temporary **feature/keyword-searches** branch.  Please let me know whether you need this capability "sooner than later".  Otherwise, I'll be considering other keywords to add besides `has_child(name)` and looking at the other two feature requests that are slated for 3.5.0.

**Comment by wwkimball** [comment]
I have published version 3.5.0, which includes this new capability.
